File: Vec3.hs
module Vec3 where

import System.Random
import Prelude hiding ((<*>))

type Vec3 = (Double, Double, Double)

(<+>) :: Vec3 -> Vec3 -> Vec3
(x0, y0, z0) <+> (x1, y1, z1) = ((x0 + x1), (y0 + y1), (z0 + z1))

(<->) :: Vec3 -> Vec3 -> Vec3
(x0, y0, z0) <-> (x1, y1, z1) = ((x0 - x1), (y0 - y1), (z0 - z1))

(<*>) :: Vec3 -> Double -> Vec3 
(x, y, z) <*> s = ((s * x), (s * y), (s * z))

(</>) :: Vec3 -> Double -> Vec3 
(x, y, z) </> s = ((x / s), (y / s), (z / s))

normalize :: Vec3 -> Vec3
normalize v =  v <*> (1.0 / sqrt (dot v v))

dot :: Vec3 -> Vec3 -> Double 
dot (x0, y0, z0) (x1, y1, z1) = x0 * x1 + y0 * y1 + z0 * z1

distance :: Vec3 -> Vec3 -> Double
distance v0 v1 = sqrt (dot (v0 <-> v1) (v0 <-> v1))

cross :: Vec3 -> Vec3 -> Vec3
cross (x0, y0, z0) (x1, y1, z1) = (y0 * z1 - z0 * y1, z0 * x1 - x0 * z1, x0 * y1 - y0 * x1)

zero :: Vec3
zero = (0.0, 0.0, 0.0)

reflect :: Vec3 -> Vec3 -> Vec3
reflect v n = v <-> (n <*> (2 * (dot v n)))

gammaCorrect :: Vec3 -> Vec3
gammaCorrect (r, g, b) = 
    let r' = min (max (sqrt r) 0.0) 1.0
        g' = min (max (sqrt g) 0.0) 1.0
        b' = min (max (sqrt b) 0.0) 1.0
    in (r', g', b')

-- http://mathworld.wolfram.com/DiskPointPicking.html
-- Returns a point in the unit sphere in the x-y plane
randomInUnitDisk :: Int -> Vec3
randomInUnitDisk seed = 
    let r0 = randomR (0.0, 1.0) (mkStdGen seed)
        r1 = randomR (0.0, 2.0 * pi) (snd r0)
        r = sqrt (fst r0)
        theta = fst r1
    in  (r * cos(theta), r * sin(theta), 0.0)

-- http://math.stackexchange.com/questions/87230/picking-random-points-in-the-volume-of-sphere-with-uniform-probability
randomInUnitSphere :: Int -> Vec3
randomInUnitSphere seed = 
    let r0 = randomR (-1.0, 1.0) (mkStdGen seed)
        r1 = randomR (-1.0, 1.0) (snd r0)
        r2 = randomR (-1.0, 1.0) (snd r1)
        r3 = randomR (0.0, 1.0) (snd r2)
        x = fst r0
        y = fst r1
        z = fst r2
        u = fst r3
    in (normalize (x, y, z)) <*> (u ** (1.0 / 3.0))

File: raytracer-cpp/src/vec3.cpp
#include "vec3.h"

Vec3 operator+(const Vec3 &u, const Vec3 &v) {
    return Vec3(u.x + v.x, u.y + v.y, u.z + v.z);
}

Vec3 operator-(const Vec3 &u, const Vec3 &v) {
    return Vec3(u.x - v.x, u.y - v.y, u.z - v.z);
}

Vec3 operator*(float s, const Vec3 &v) {
    return Vec3(s * v.x, s * v.y, s * v.z);
}

Vec3 Vec3::normalize(const Vec3 &v) {
    float l = Vec3::length(v);

    if (l == 0.0) {
        return v;
    }

    return (1.0 / l) * v;
}

float Vec3::length(const Vec3 &v) {
    return sqrt(v.x * v.x + v.y * v.y + v.z * v.z);
}

Vec3 Vec3::reflect(const Vec3 &v, const Vec3 &n) {
    return v - (2.0 * Vec3::dot(v, n)) * n;
}

float Vec3::dot(const Vec3 &v1, const Vec3 &v2) {
    return v1.x * v2.x + v1.y * v2.y + v1.z * v2.z;
}

Vec3 Vec3::cross(const Vec3 &v1, const Vec3 &v2) {
    float x = v1.y * v2.z - v1.z * v2.y;
    float y = v1.z * v2.x - v1.x * v2.z;
    float z = v1.x * v2.y - v1.y * v2.x;
    return Vec3(x, y, z);
}

// http://mathworld.wolfram.com/DiskPointPicking.html
Vec3 Vec3::random_in_unit_disk() {
    float r = RAND(0.0, 1.0);     
    float theta = RAND(0.0, 2.0 * M_PI);
    return Vec3(r * cos(theta), r * sin(theta), 0.0);
}

// http://math.stackexchange.com/questions/87230/picking-random-points-in-the-volume-of-sphere-with-uniform-probability
Vec3 Vec3::random_in_unit_sphere() {
    float x = RAND(-1.0, 1.0);
    float y = RAND(-1.0, 1.0);
    float z = RAND(-1.0, 1.0);
    float u = RAND(0.0, 1.0);
    return pow(u, 1.0 / 3.0) * Vec3::normalize(Vec3(x, y, z));
}

Vec3 Vec3::clamp(const Vec3 &v, float min, float max) {
    Vec3 result = v;

    if (result.x < min) {
        result.x = min;
    }
    if (result.x > max) {
        result.x = max;
    }

    if (result.y < min) {
        result.y = min;
    }
    if (result.y > max) {
        result.y = max;
    }

    if (result.z < min) {
        result.z = min;
    }
    if (result.z > max) {
        result.z = max;
    }

    return result;
}

Vec3 Vec3::rotate_x(const Vec3 &v, float cos_theta, float sin_theta) {
    return Vec3(v.x, cos_theta * v.y - sin_theta * v.z, sin_theta * v.y + cos_theta * v.z);
}

Vec3 Vec3::rotate_y(const Vec3 &v, float cos_theta, float sin_theta) {
    return Vec3(cos_theta * v.x + sin_theta * v.z, v.y, -sin_theta * v.x + cos_theta * v.z);
}

Vec3 Vec3::rotate_z(const Vec3 &v, float cos_theta, float sin_theta) {
    return Vec3(cos_theta * v.x - sin_theta * v.y, sin_theta * v.x + cos_theta * v.y, v.z);
}
